### ISSUE TITLE
CP15 barrier instructions should be emitted before the exclusives loops (arm)

### ISSUE THREAD

**Issue by zrzka** [rust-lang/rust#60605]
## Symptoms

* Rustup hangs (unable to install Rust)
  * Tracked down to [parking_lot](https://github.com/Amanieu/parking_lot/issues/130) issue
* Not just Rust related, [Swift has this problem as well](https://forums.balena.io/t/cloud-build-fails-but-local-device-build-works-on-raspberry-pi-zero/4994)
  * -> LLVM issue

## Environment

* Linux kernel with`CP15_BARRIER_EMULATION=y`
* `abi.cp15_barrier` set to `1` (emulate)
* `arm-unknown-linux-gnueabihf` toolchain

## CP15 barrier instructions

* They're deprecated since `armv7`
* Linux kernel can [emulate or HW exec them](https://www.kernel.org/doc/Documentation/arm64/legacy_instructions.txt)
  * `abi.cp15_barrier` is set to `2` (HW exec) -> there's no issue
    * The CPU must support them
    * ARMv8 in our case, which still supports them
  * `abi.cp15_barrier` is set to `1` (emulate) -> there's this issue

## Issue description

`parking_lot` author:

> This seems to be closer to an LLVM bug than a parking_lot bug. The source of the problem is the CP15 emulation in the kernel. Essentially the `mcr p15, #0x0, r12, c7, c10, #0x5` is trapping to the kernel every time, which invalidates the exclusive monitor between the `ldrex` and `strex` instructions. This results in the `strex` never succeeding and looping indefinitely.

<img width="1184" alt="instructions-loop" src="https://user-images.githubusercontent.com/1084172/57302722-1e997400-70dc-11e9-968c-7f1e8dc0595b.png">

ARM engineer (Will Deacon) response on this:

> Hi again, Robert,
> 
> Just a quick update on this:
>
>  1. CP15 barriers remain deprecated in the Armv8 architecture, and so
>    may be removed entirely from future CPUs.
>
> 2. Because of (1), the kernel defaults to trap+emulate, so that it can
>    warn about the use of these instructions. I think this is the right
>    thing to do because, once the instructions have been removed, we
>    will have no choice but to trap+emulate (this happened for the SWP
>    instruction already). This trapping will prevent your exclusives loop
>    from ever succeeding.
>
> 3. The right place to address this issue is in LLVM, where atomic
>    read-modify-write operations with conditional release semantics (i.e.
>    release on success) should actually emit the CP15 barrier before the
>    exclusives loop. Assuming that contention is rare (which it kind of
>    needs to be for performant compare-and-swap anyway), I don't see this
>    having a meaningful impact on performance.
>
> I've reached out to one of our upstream LLVM developers, and I'll be talking
> with him face-to-face next week about getting this fixed.
>
> Will

## Solution

Will's third point:

> Atomic read-modify-write operations with conditional release semantics (i.e.
> release on success) should actually emit the CP15 barrier before the
> exclusives loop. Assuming that contention is rare (which it kind of
> needs to be for performant compare-and-swap anyway), I don't see this
> having a meaningful impact on performance.

And:

> I've reached out to one of our upstream LLVM developers, and I'll be talking
> with him face-to-face next week about getting this fixed.

I asked for the LLVM bug # to track it, but still no response.

## Way forward

* Fix it on the Rust LLVM fork
* Wait for the LLVM to have this fixed and wait till Rust's fork synces

Second way can prolong fix by weeks, months, ...? Not sure how fast is the LLVM itself developed & how fast is the Rust's fork syncing. This is the main reason I did report it here as well.

## No fix

People aren't / won't be able to use Rust on Linux with `CP15_BARRIER_EMULATION=y` & `abi.cp15_barrier=1` (emulation, default value) & `arm-unknown-linux-gnueabihf` toolchain.

**Comment by zrzka** [comment]
In case anyone will have this issues as well ...

## Rust installation workaround

Force installer to use `curl` instead of `reqwest` crate.

```
export RUSTUP_USE_CURL=1
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

**NOTE:** This workaround allows you to install Rust toolchain only, but it doesn't fix the issue. Compile the example below, run it and you'll see.

## Example which reproduces this problem

`Cargo.toml`:

```
[package]
name = "cp15"
version = "0.1.0"
edition = "2018"

[dependencies]
reqwest = "0.9"
```

`src/main.rs`:

```rust-lang
fn main() {
    let body = reqwest::get("https://www.rust-lang.org");
    println!("Response: {:?}", body);
}
```

Compile with `arm-unknown-linux-gnueabihf` and run on the kernel with `CP15_BARRIER_EMULATION=y` & `abi.cp15_barrier=1` => infinite loop, never ends.

**Comment by nagisa** [comment]
Very nicely investigated issue. 

> Fix it on the Rust LLVM fork

Fixing this in our fork involves knowing what the fix should look like. I imagine the easiest and fastest approach will be to wait for LLVM to get a patch from one of the ARM’s engineers and then include that patch in our fork.

**Comment by zrzka** [comment]
Agree. Will ping Will again to get the LLVM bug # (if exists) and the patch which will fix this.

**Comment by zrzka** [comment]
Did report it in the LLVM's Bugzilla: https://bugs.llvm.org/show_bug.cgi?id=41856

**Comment by comex** [comment]
It occurs to me that even once this bug is fixed, making every compare-and-swap (or at least every one using those semantics) trap to the kernel is going to be rather slow.  The non-deprecated equivalent of the CP15 barrier is the `DMB` instruction, but that's only supported starting with ARMv7, whereas Rust's `arm-unknown-linux-gnueabihf` target only assumes ARMv6.  However, Rust also has `armv7-unknown-linux-gnueabihf` which should be ABI compatible and avoid the issue.  Since the cp15 trapping apparently only exists on arm64 kernels, `aarch64-unknown-linux-gnu` is also worth a mention.  Both `armv7` and `aarch64` variants are tier 2, same as the `arm` variant, so I guess there's not much for Rust to do here.

But I am curious why rustup is picking `arm`, since it identifies architecture based on `uname -m`, which should reflect the kernel architecture.  Even if you're running a 32-bit userland and a 64-bit kernel, it should still show aarch64.  And if you're running a 32-bit kernel, why do you have `cp15_barrier`, whose implementation lives under `arch/arm64`?...

**Comment by zrzka** [comment]
@comex the short answer is _cross compilation_ (in quotes).

> But I am curious why rustup is picking `arm`.

[Our](https://www.balena.io) customers have dozens of different devices - RPi Zero, RPi 1, RPi 3, ... These device are ARMv6, v7, ... Part of our infrastructure is a service called builder. You can push to a special git origin and your service is going to be build on our builder. Building on RPi Zero / via QEMU is slow. Thus our build servers have ARMv8 CPUs. We're using _tricks_ like [personality](http://man7.org/linux/man-pages/man2/personality.2.html), [setarch](http://man7.org/linux/man-pages/man8/setarch.8.html), ability to run 32-bit user space on 64-bit kernel, docker, ... Which basically solves couple of issues like:

* builds are fast,
* same toolchain is used as on real HW,
* same OS variant is used (like `armhf` Debian for RPi Zero),
* ...

We solved our problem by setting `abi.cp15_barrier` to `2`, which means HW exec. But in the future, when ARMv9 will be released, ... and these instructions will be removed (probably), we will have to set `abi.cp15_barrier` to `1` and then all these _tricks_ won't work.

**Comment by MichaIng** [comment]
Faced the same on ARMv8 platform with 64-bit kernel but 32-bit userland when enforcing `armv7-unknown-linux-gnueabihf` host triple. Many thanks for providing the workaround `export RUSTUP_USE_CURL=1` 👍, I was lost until finding this.

**Comment by Jamesbarford** [comment]
Triage; this can be re-created with the following code, essentially a Rust repro of; https://github.com/llvm/llvm-project/issues/41201

```rust
// repro.rs

use std::sync::atomic::{AtomicI32, Ordering};

#[no_mangle]
fn cas_release(p: &AtomicI32, o: i32, n: i32) -> i32 {
    loop {
        let e = o;
        match p.compare_exchange_weak(
            e,
            n,
            Ordering::Release,
            Ordering::Relaxed,
        ) {
            Ok(_) => return 0,
            Err(current) if current != o => return -1,
            Err(_) => continue,
        }
    }
}

#[no_mangle]
fn cas_acq_rel(p: &AtomicI32, o: i32, n: i32) -> i32 {
    loop {
        let e = o;
        match p.compare_exchange_weak(
            e,
            n,
            Ordering::AcqRel,
            Ordering::Relaxed,
        ) {
            Ok(_) => return 0,
            Err(current) if current != o => return -1,
            Err(_) => continue,
        }
    }
}

fn main() {}
```

compile with; `rustc --target=arm-unknown-linux-gnueabihf -O ./repro.rs`


Which emits;
```asm
ldrex   r3, [r0]
cmp     r3, r1
bne     .LBB6_1
mcr     p15, #0, lr, c7, c10, #5
strex   r4, r2, [r0]
cmp     r4, #0
```
